File: api/_shared/rate_limit.py
"""Best-effort per-IP rate limiting for the /api inference functions.

Each of these endpoints can trigger multiple XGBoost model calls plus live
weather/sensor lookups per request — cheap to hammer, not cheap to serve.
Vercel's Fluid Compute reuses warm function instances across concurrent
requests (see Context/Chunks/heatroute.md), so this in-memory counter is not
a no-op the way it would be under classic one-request-per-instance
serverless — it actually throttles repeat callers on the same warm instance.
It resets on a cold start and is not shared across instances, so it is a
mitigation against casual abuse/cost spikes, not a hard security boundary —
add a real edge-level limiter (e.g. Vercel Firewall) before this matters for
a production launch.
"""

from __future__ import annotations

import time

_WINDOW_SECONDS = 60.0
_MAX_REQUESTS_PER_WINDOW = 30

# client_ip -> (window_start_epoch_s, count_in_window)
_buckets: dict[str, tuple[float, int]] = {}


def client_ip(headers) -> str:
    forwarded = headers.get("X-Forwarded-For")
    if forwarded:
        return forwarded.split(",")[0].strip()
    return headers.get("X-Real-IP", "unknown")


def check(headers) -> bool:
    """Returns True if the request should proceed, False if it should be
    rejected (429). Prunes old buckets opportunistically so this doesn't
    grow unbounded on a long-lived warm instance."""
    ip = client_ip(headers)
    now = time.time()
    if len(_buckets) > 5000:
        _buckets.clear()

    window_start, count = _buckets.get(ip, (now, 0))
    if now - window_start >= _WINDOW_SECONDS:
        window_start, count = now, 0

    count += 1
    _buckets[ip] = (window_start, count)
    return count <= _MAX_REQUESTS_PER_WINDOW

File: src/lib/rate-limit.ts
// Best-effort per-IP rate limiting for auth Server Actions, on top of
// Supabase's own auth.rate_limit config — this catches obvious scripted
// abuse at the app layer before it even reaches Supabase. Resets on a cold
// server restart and isn't shared across instances, so it's a mitigation,
// not a hard boundary (see api/_shared/rate_limit.py for the same trade-off
// on the Python inference functions).
const WINDOW_MS = 5 * 60_000;
const MAX_PER_WINDOW = 10;

const buckets = new Map<string, { windowStart: number; count: number }>();

export function isRateLimited(key: string): boolean {
  const now = Date.now();
  if (buckets.size > 5000) buckets.clear();

  const bucket = buckets.get(key);
  if (!bucket || now - bucket.windowStart >= WINDOW_MS) {
    buckets.set(key, { windowStart: now, count: 1 });
    return false;
  }
  bucket.count += 1;
  return bucket.count > MAX_PER_WINDOW;
}
